[bugfix-627] build BSM with deb and flatpak for linux

silentrald opened the pull request:
Closes #627, Closes #468

- [x] Add docs for installing the different BSM builds

### NOTES

- Already tested the `npm run publish:linux` and it should work on the ubuntu-latest
- Docs assumes that #586 is already merged

Zagrios commented:
I will review the #586 tomorow to merge it before this one
When this one is merge I will probably release a alpha.5

Zagrios commented:
@silentrald is it relevant to add more than the `.deb` ? From what I see here Steam is only available throught a `.deb` as well :thinking: 
https://github.com/ValveSoftware/steam-for-linux

silentrald commented:
@Zagrios I think some `deb` installers are supported but not recommended for all distros.

In Arch Linux's case, we have a different package manager `pacman` which is different with `apt` so the problem here is that if we install `deb` then either we have missing/incompatible dependencies or manage 2 package manager (pacman and apt) which is not ideal.

https://wiki.archlinux.org/title/Creating_packages_for_other_distributions#Tips_and_Tricks_about_Debian

I think same is true with Fedora that it would prefer to install `rpm` rather than `deb` because that is native to their distro as well. Although I don't have a wiki/article to link but there are many sentiments/opinions by Fedora users when you try searching on the net.

So building it with different targets would be best, compatibility wise. The only question now is that, is there any drawback that you probably see into building it into different targets? Except for the build time at least.

Insprill commented on docs/wiki/Linux.md:
```diff
@@ -0,0 +1,55 @@
+
+Once Proton is installed, when you open your BSManager application for the first time, it will ask you to link the _Proton Folder_ so that it can access the `proton` binary inside it. This will also check if the folder you've selected is valid or not. You can also change this in the future by going to the **settings page** and search for the _Proton Folder_.
+
+## Wine Install [Optional]
+
+[Wine](https://www.winehq.org/) is a tool to run windows application under Linux. BSManager uses to run [BSIPA](https://nike4613.github.io/BeatSaber-IPA-Reloaded/) so that you can play Beat Saber with mods. You can install `wine` depending on your package manager. If you don't install it, this will just fallback to using the `wine` executable found in your _Proton Folder_.
```
Is this section still required now that we use wine from proton?

silentrald replied:
Yep, just forgot to update, should be in commit [38335f1](https://github.com/Zagrios/bs-manager/pull/629/commits/38335f1f44db358a8f322b941e0ce0e4abe5bcf1)

Zagrios commented:
@silentrald is this PR ready ?
What about the flatpak thing ?

silentrald commented:
@Zagrios might get a commit flatpak by today. Since it's universal, do you think we need to drop support on any installer or leave them as is so users can still have a decision of what to install?

Zagrios commented:
@silentrald  I think we might keep at least the `.deb` and maybe the `.rpm` just in case 👌 
Take your time, I won’t be very available this weekend to review any PRs anyway 😉

Insprill commented:
I'd be fine dropping the pacman package. Afaik we can't really customize it that much, and anyone on Arch is going to use the AUR anyway. The AUR package(s) will either be built from source where this doesn't matter at all or we can simply extract the `.deb`.

silentrald commented:
Added flatpak and dropped rpm and pacman.

I've already tested the github action build release and it should be working, you can verify this from my fork https://github.com/silentrald/bs-manager/actions/runs/11752857569

---

@Zagrios Also as a side tangent as well, if its ok for both me and @Insprill to be added in the README.md file for the credits with the linux contribution?

ltsymes23 commented on docs/wiki/Linux.md:
````diff
@@ -0,0 +1,61 @@
+You are required to have `flatpak` installed your system. After installing that, download the `.flatpak` file in the releases and run the following command:
+
+```bash
+flatpak install ./bsmanager.flatpak
````
Potential change to avoid permission issues:
flatpak install --user ./bsmanager.flatpak

ltsymes23 commented on docs/wiki/Linux.md:
```diff
@@ -0,0 +1,61 @@
+
+### Universal (flatpak)
+
+You are required to have `flatpak` installed your system. After installing that, download the `.flatpak` file in the releases and run the following command:
```
General idiot-proofing the instructions:

You are required to have `flatpak` installed your system, if it is not installed then do so using your system's package manager ("sudo apt install flatpak" for Ubuntu, or "sudo pacman -S flatpak" for Arch). After installing, download the `.flatpak` file from the releases, CD to the download location, and run the following command:

Insprill replied:
Is there really a case where someone wants to use the flatpak instead of the AUR package / deb, and they don't have it installed, and don't know how to install it? I could be wrong here, but it seems redundant to me.

silentrald replied:
Yeah, I think other distro's should be listed here like [Fedora](https://developer.fedoraproject.org/deployment/flatpak/flatpak-install.html) and [NixOS](https://flatpak.org/setup/NixOS) which isn't Debian or AUR supported.

ltsymes23 commented on docs/wiki/Linux.md:
````diff
@@ -0,0 +1,61 @@
+
+chown $(whoami) /opt/BSManager/resources/assets/jsons/bs-versions.json
+```
+
````
(Editors Note: the following info was taken from https://discord.com/channels/1049624409276694588/1281752371340513403 and should be double checked before publishing)


Steam Deck instructions:

The Steam Deck has a read-only filesystem by default so extra steps are needed:


Before running other commands, disable read-only mode:

sudo steamos-readonly disable


Then do the following before proceeding with the desired install method above:

sudo pacman-key --init

sudo pacman-key --populate holo

sudo pacman -S screen

sudo pacman -S dotnet-sdk


Once BSManager is installed, re-enable read-only mode:

sudo steamos-readonly enable

Insprill replied:
Those instructions are outdated as of 586 being merged; you no longer need any system dependencies.

Lohkdesgds commented:
Man I may be crazy but I am testing that deb app there just for fun
I won't be reviewing even though I want this so bad. Someone that knows more about this code can do that I think :P 

Thank you for making it work on linux ❤️

Zagrios commented:
> Added flatpak and dropped rpm and pacman.
> 
> I've already tested the github action build release and it should be working, you can verify this from my fork https://github.com/silentrald/bs-manager/actions/runs/11752857569
> 
> @Zagrios Also as a side tangent as well, if its ok for both me and @Insprill to be added in the README.md file for the credits with the linux contribution?

Of course 👌 Add your names to the Readme in this PR 😉

silentrald commented:
> Of course 👌 Add your names to the Readme in this PR 😉

@Zagrios Thank you, will create in the next commit.

Although I'm still having problems with the flatpak build though, I confirmed with @LiamillionSS that you can run the flatpak of BSM but you can't run any BS version because of some flatpak sandboxing rules. I'm still investigating/researching it, though I'm worrying if I can push it to the v1.5.0 release. If I can't meet the release deadline, I might commit a push to temporarily support AppImage for now as the universal linux release and I can support with the deep linking configuration. What do you think?

Zagrios commented:
@silentrald I saw your update message. But yeah if we can't have proper a working solution with Flatpak, we can first release only the `.deb` and `.rpm` files, and then release a `.flatpak` later 👌

Zagrios commented:
@silentrald I've pushed a small change to be able to package and publish on Windows.

But I've downloaded the build artifact from your repo action, and I get this error when I try to open BSM after installing the Flatpak: `Failed to sync with dbus proxy`

I couldn't launch the `.deb` either, but I believe that's because I already have a BSM instance installed from a `.deb` :thinking:

silentrald commented:
@Zagrios

> But I've downloaded the build artifact from your repo action, and I get this error when I try to open BSM after installing the Flatpak: Failed to sync with dbus proxy

What was your installation process/steps? Did you follow the installation guide in the *Linux.md* file in this commit as well?

> I couldn't launch the .deb either, but I believe that's because I already have a BSM instance installed from a .deb 🤔

You could do a
```
sudo dpkg --remove bs-manager
```
but the dpkg install should automatically overwrite the files on `dpkg install` afaik 🤔. Or there are conflicting files within your system that isn't deb, you might need to manually delete/remove the files in most of the `/opt/BSManager` directory in any case.

Zagrios commented:
@silentrald 
Okay, the `.flatpak` is working; it's just that right after the installation, I still get the error. However, starting BSManager from the explorer works :shrug:

https://github.com/user-attachments/assets/69f646bc-f691-4667-9360-2de2f0e4c30c

For the `.deb`, it didn’t work directly. I need to start BSM directly from the `/opt/BSManager` directory or through the terminal with the `--no-sandbox` argument.  
Otherwise, I get this error:
```
The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /opt/BSManager/chrome-sandbox is owned by root and has mode 4755.
Trace/breakpoint trap (core dumped)
```


https://github.com/user-attachments/assets/b22bb07d-fdbc-4276-8898-935cf37c791a

Zagrios commented:
> @silentrald Okay, the `.flatpak` is working; it's just that right after the installation, I still get the error. However, starting BSManager from the explorer works 🤷
> 2024-11-19.20-21-22.mp4
> 
> For the `.deb`, it didn’t work directly. I need to start BSM directly from the `/opt/BSManager` directory or through the terminal with the `--no-sandbox` argument. Otherwise, I get this error:
> 
> ```
> The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /opt/BSManager/chrome-sandbox is owned by root and has mode 4755.
> Trace/breakpoint trap (core dumped)
> ```
> 
> 2024-11-19.20-40-15.mp4

From what I see, Signal had the same issue with the sandbox thing with the `.deb`.  
They had to patch it manually here: 
https://github.com/signalapp/Signal-Desktop/blob/0bc6368c642a7ddf2456e994db1016034380f72d/patches/app-builder-lib%2B24.13.3.patch#L149 

I will not have the time to look into it further tonight unfortunately :cry:

silentrald commented:
@Zagrios 

> silentrald Okay, the `.flatpak` is working; it's just that right after the installation, I still get the error. However, starting BSManager from the explorer works 🤷

I think we can just give the installation process documented in the *Linux.md* file tbh and we can worry about this in the future if we want to release the application in flathub. Might need to check it another time on how this works.

You might also want to test whether you can launch the BS executable if it works on your end as well. Check if mods are injected properly as well.
 
> For the `.deb`, it didn’t work directly. I need to start BSM directly from the `/opt/BSManager` directory or through the terminal with the `--no-sandbox` argument. Otherwise, I get this error:
> 
> ```
> The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /opt/BSManager/chrome-sandbox is owned by root and has mode 4755.
> Trace/breakpoint trap (core dumped)
> ```

When I was developing this, I had to edit my permissions due to the update to Ubuntu 24.04 as documented here https://github.com/electron/electron/issues/42510#issuecomment-2171583086. But looking into the error message itself and within the issue in electron says as well, we could just change the permissions of `/opt/BSManager/chrome-sandbox` to 4755. I just have to add it in the `build/after-install.sh`.

To verify if the installation succeeded with the correct permissions, you can do this:
```bash
ls -la /opt/BSManager | grep "chrome-sandbox"
# expects: -rwsr-xr-x 1 root root ... chrome-sandbox
```

silentrald commented:
@Zagrios got every actions on the green. Everything should work as expected now, please retest the mods installation and game launching for both linux and windows. Build versions (nsis, deb and flatpak) needs to be verified as well ig. So that if there are any issues encountered, I can fix them immediately.

Zagrios commented:
@silentrald 
### Issues with flatpak:
- When enabling oneclicks, they stay disabled
- Cannot install maps to original Steam version (because of sandbox i guess ?) (same with models and mods):
```
[error] Error: ENOENT: no such file or directory, mkdir '/home/mathieu/.local/share/Steam/steamapps/common/Beat Saber/Beat Saber_Data/CustomLevels/41d68 (unending thoughts of everyone and everything - member63432_19, member63432_63, member63432_-210^3)' ENOENT undefined
```
- I can't launch any BeatSaber version (original or downgraded):
```
[2024-11-22 20:53:41.933] [info]  Linux exec command
> flatpak-spawn --host  ps awwxo pid,args | grep "steam-runtime-launcher-service"
[2024-11-22 20:53:41.963] [error] Error while checking if Steam is running as admin Error: Unable to build process from pid: File not found: /proc/15251
    at p.isElevated (/app/lib/org.erb.BSManager/resources/app.asar/dist/main/main.js:14:1217660)

[2024-11-22 20:53:41.964] [info]  Linux exec command
> flatpak-spawn --host  ps awwxo args | grep -c "steam-runtime-launcher-service"
[2024-11-22 20:53:41.990] [info]  Linux spawn command
> flatpak-spawn --host --env=SteamAppId="620980" --env=SteamOverlayGameId="620980" --env=SteamGameId="620980" --env=WINEDLLOVERRIDES="winhttp=n,b" --env=STEAM_COMPAT_DATA_PATH="/home/mathieu/.steam/steam/steamapps/compatdata/620980" --env=STEAM_COMPAT_INSTALL_PATH="/home/mathieu/.local/share/Steam/steamapps/common/Beat Saber" --env=STEAM_COMPAT_CLIENT_INSTALL_PATH="/home/mathieu/.steam/steam" --env=STEAM_COMPAT_APP_ID="620980" --env=SteamEnv="1" "/home/mathieu/.steam/steam/steamapps/common/Proton - Experimental/proton" run "/home/mathieu/.local/share/Steam/steamapps/common/Beat Saber/Beat Saber.exe" --no-yeet fpfc
[2024-11-22 20:53:52.005] [error] BS process unref after timeout 10000

```


### With the deb:
- BSM ask for my keychain for some reason ?
- Can't launch BS version >= 1.39.1 (version below that works):
![image](https://github.com/user-attachments/assets/a0acae07-86e4-49d6-b7ac-0c4d21d66be9)
- Installing BSIPA produce tons of errors, but it does install :sweat_smile: (this is just a sample) :
```
[2024-11-22 21:08:46.173] [error] IPA process stderr 0024:err:mscoree:LoadLibraryShim error reading registry key for installroot

[2024-11-22 21:08:46.177] [error] IPA process stderr 0024:err:mscoree:LoadLibraryShim error reading registry key for installroot

[2024-11-22 21:08:46.219] [error] IPA process stderr 0024:fixme:nls:GetFileMUIPath stub: 0x10, L"C:\\windows\\system32\\tzres.dll", (null), 000000000011E7C0, 0000000000944CE0, 000000000011E7C8, 000000000011E7B8
```
- I cannot disable OneClicks (But I believe that's exptected ? :thinking:)



*For some reason, after testing the `.deb` I can launch BeatSaber, install maps models and mods with the `.flatpak`; I will reboot my PC  to see if that change something*

silentrald commented:
@Zagrios 

- For the OneClick settings, the electron js code doesn't work with Linux. Also created an #651 issue to delay the fix since I have to check if the fix is possible.

- For the flatpak install, yeah this needs to have create/rw permission to the Beat Saber folder so just needs to add this
```
~/.steam/steam/steamapps/common/Beat Saber:create
```

- I could run 1.39.1 version for both deb and flatpak 🤔. Can you reinstall this in any case it's just that specific version you we're testing was corrupted/broken?

- BSIPA installs with wine throw a lot or unnecessary stderr but as long as the install works, it shouldn't be a problem. It's just good to have logs if any errors would pop out in users BSIPA installations.

- The keyring, this one I'm not too keen what really it is. But it has something to do with logging in, if you skip the ubuntu login page. Although I have to check if this is a system thing or just the application itself asking to get unecrypted (even if I don't add a password it still runs fine 🙃)

silentrald commented:
@Zagrios The keyring seems to be a system thing with the user either needs to login or disable it from their machine. Ive also checked on electron-builder configs, seems theres nothing there and also in apt/dpkg docs doesn't seem to have anything relating to this. So probably safe to ignore this.

silentrald commented:
@Zagrios got it. Release Linux should be this one https://github.com/silentrald/bs-manager/actions/runs/12021117904/job/33511017646

Zagrios commented:
@silentrald 

### Flatpak
- Now i can install things to my original copy of Beat Saber :ok_hand: 
- Now i can't launch any copy of Beat Saber :sweat_smile: (I've tried to reboot but nothing change)
```
[2024-11-28 20:57:52.014] [info]  Linux exec command
> flatpak-spawn --host  ps awwxo args | grep -c "steam-runtime-launcher-service"
[2024-11-28 20:57:52.034] [error] Error: Command failed: flatpak-spawn --host  ps awwxo args | grep -c "steam-runtime-launcher-service"
Portal call failed: org.freedesktop.DBus.Error.ServiceUnknown

    at genericNodeError (node:internal/errors:984:15)
    at wrappedFn (node:internal/errors:538:14)
    at ChildProcess.exithandler (node:child_process:422:12)
    at ChildProcess.emit (node:events:519:28)
    at maybeClose (node:internal/child_process:1105:16)
    at Socket.<anonymous> (node:internal/child_process:457:11)
    at Socket.emit (node:events:519:28)
    at Pipe.<anonymous> (node:net:339:12)
[2024-11-28 20:57:55.964] [error] Unable to open steam
[2024-11-28 20:57:55.965] [info]  Linux spawn command
> flatpak-spawn --host --env=SteamAppId="620980" --env=SteamOverlayGameId="620980" --env=SteamGameId="620980" --env=WINEDLLOVERRIDES="winhttp=n,b" --env=STEAM_COMPAT_DATA_PATH="/home/mathieu/.steam/steam/steamapps/compatdata/620980" --env=STEAM_COMPAT_INSTALL_PATH="/home/mathieu/.local/share/Steam/steamapps/common/Beat Saber" --env=STEAM_COMPAT_CLIENT_INSTALL_PATH="/home/mathieu/.steam/steam" --env=STEAM_COMPAT_APP_ID="620980" --env=SteamEnv="1" "/home/mathieu/.steam/steam/steamapps/common/Proton - Experimental/proton" run "/home/mathieu/.local/share/Steam/steamapps/common/Beat Saber/Beat Saber.exe" fpfc
[2024-11-28 20:57:55.984] [info]  BS process exit with code 1
[2024-11-28 20:57:55.984] [info]  BS process exit code 1

```
- BSM has lost its icon :cry: (And it should be `BSManager` not `bs-manager`)
![image](https://github.com/user-attachments/assets/1bc059c7-9315-483c-8e29-4c574d18d8f9)
- Cannot downgrade BS versions (same issue as the .deb):
```
[2024-11-28 21:22:23.169] [error] Error: spawn /app/lib/com.electron.bs_manager/resources/assets/scripts/DepotDownloader ENOENT
    at Process.ChildProcess._handle.onexit (node:internal/child_process:286:19)
    at onErrorNT (node:internal/child_process:484:16)
    at processTicksAndRejections (node:internal/process/task_queues:82:21) ENOENT undefined
```



### Deb:
- Also lost its icon (rip)
- Can launch Beat Saber
- Cannot downgrade BS versions
```
[2024-11-28 21:13:03.006] [error] Error: spawn /opt/bs-manager/resources/assets/scripts/DepotDownloader ENOENT
    at Process.ChildProcess._handle.onexit (node:internal/child_process:286:19)
    at onErrorNT (node:internal/child_process:484:16)
    at processTicksAndRejections (node:internal/process/task_queues:82:21) ENOENT undefined
```
![image](https://github.com/user-attachments/assets/6b042cfa-c87c-4d57-9296-a8c263d499d6)

silentrald commented:
@Zagrios

> Now i can't launch any copy of Beat Saber 😅 (I've tried to reboot but nothing change)
This is related to permissions. I've already added that in the default but you could try to reset the permissions using FlatSeal. When you reset it you should see `org.freedesktop.Flatpak`.

[Checking the DepotDownloader and Icons Issue]

Zagrios commented:
@silentrald Great news, everything seems to work! 🎉
Can you confirm that everything is ready to merge? :eyes:

silentrald commented:
> silentrald Great news, everything seems to work! 🎉
> Can you confirm that everything is ready to merge? :eyes: 

@Zagrios confirmed 👌

Zagrios commented:
Merging 🚀 🚀
